File: 4257-sum-of-sortable-integers/4257-sum-of-sortable-integers.java
class Solution {
    public int sortableIntegers(int[] nums) {
        int n = nums.length;
        ArrayList<Integer> factors = new ArrayList<>();
        factors = getFactors(n);
        int res = 0;
        for (int ele : factors) {
            if (check(ele, nums)) res += ele;
        }
        return res;
    }

    private boolean check(int k, int arr[]) {
        int n = arr.length;
        int ranges[][] = new int[n / k][2];
        int idx = 0, curr = 0;
        while (curr < n) {
            ranges[idx][0] = curr;
            ranges[idx][1] = curr + k - 1;
            idx++;
            curr = curr + k;
        }
        int prevMax = Integer.MIN_VALUE;
        for (int i = 0; i < ranges.length; i++) {
            int currLeft = ranges[i][0], currRight = ranges[i][1];
            int currMin = Integer.MAX_VALUE, currMax = Integer.MIN_VALUE;
            int drop = 0;
            for (int j = 0; j < k; j++) {
                int val = arr[currLeft + j];
                currMin = Math.min(currMin, val);
                currMax = Math.max(currMax, val);
                int nextVal = arr[currLeft + ((j + 1) % k)];
                if (val > nextVal) 
                    drop++;
            }
            if (drop > 1) return false;
            if (i > 0) {
                if (currMin < prevMax) return false;
            }
            prevMax = currMax;
        }
        return true;
    }

    private ArrayList<Integer> getFactors(int num) {
        ArrayList<Integer> res = new ArrayList<>();
        for (int i = 1; i * i <= num; i++) {
            if (num % i == 0) {
                res.add(i);
                if (num / i != i) {
                    res.add(num / i);
                }
            }
        }
        return res;
    }
}

File: 4257-sum-of-sortable-integers/4257-sum-of-sortable-integers.py
class Solution:
    def sortableIntegers(self, nums):
        n = len(nums)
        factors = self.getFactors(n)
        res = 0

        for k in factors:
            if self.check(k, nums):
                res += k

        return res

    def check(self, k, arr):
        n = len(arr)
        prev_max = float('-inf')

        for start in range(0, n, k):
            curr_min = float('inf')
            curr_max = float('-inf')
            drop = 0

            for j in range(k):
                val = arr[start + j]
                curr_min = min(curr_min, val)
                curr_max = max(curr_max, val)

                next_val = arr[start + (j + 1) % k]
                if val > next_val:
                    drop += 1

            if drop > 1:
                return False

            if start > 0 and curr_min < prev_max:
                return False

            prev_max = curr_max

        return True

    def getFactors(self, num):
        res = []
        i = 1
        while i * i <= num:
            if num % i == 0:
                res.append(i)
                if num // i != i:
                    res.append(num // i)
            i += 1
        return res